### PR TITLE
nixos/starship: init

### PR DESCRIPTION
**Pull request by danth**
###### Motivation for this change

Added a NixOS module which configures [Starship](https://starship.rs/) for the Bash, Fish and Zsh shells.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by danth** [comment]
@bbigras @davidtwco @Br1ght0ne @Frostman @marsam
As you are maintainers of the Starship package, should I add you to `meta.maintainers`  for this module as well?

**Review by Br1ght0ne — APPROVED** [review]
LGTM.

**Comment by danth** [comment]
Rebased on master to fix merge conflicts.

File: nixos/modules/programs/starship.nix
Comment (Artturin) [inline nixos/modules/programs/starship.nix:20]:
```suggestion
    enable = mkEnableOption "Starship the shell prompt";
```
`Whether to enable` is automatically prefixed when using mkEnableOption and `.` is automatically added

**Review by Artturin — CHANGES_REQUESTED** [review]

**Comment by teto** [comment]
(just FYI, there is a package in home-manager for starship)

**Comment by danth** [comment]
Again just a rebase to fix conflicts.

**Comment by pennae** [comment]
@ofborg test starship

**Comment by pennae** [comment]
module looks good and works, but the test doesn't work right.

**Comment by danth** [comment]
@pennae

I don't particularly like how this test is written. I couldn't find a way to run each shell interactively using `machine.execute`, while being able to check its output for the presence of a Starship prompt, so I resorted to having the commands typed into a TTY. In order to keep each shell's output separate, I had to use `script` to record it to a file which is then read by `grep` after the shell is closed. It seems that the file is not always fully written before the `grep` command runs, so the test sometimes fails because the prompt was not seen.

I could split this into three separate tests, one for each shell, with the test shell set as default so that it appears after login. Then the TTY can be directly scanned for Starship's prompt. That would also have the benefit of testing each shell's configuration in isolation. I originally avoided this because I thought running three separate VM tests just for Starship might use a lot of computational resources.

What would you recommend?

**Comment by pennae** [comment]
having one test for all three shells sounds fine. maybe you can run shells with `-i` to force them into being interactive and run each one through an `expect` script to do the interacting and response checking. seems like the perfect application for that kind of thing :)

**Review by pennae — APPROVED** [review]
test works now, looks good!

File: nixos/tests/starship.nix
Comment (marsam) [inline nixos/tests/starship.nix:22]: Please use `installCheckPhase` to test this. See https://github.com/NixOS/nixpkgs/issues/143681

Besides that, LGTM

**Review by marsam — DISMISSED** [review]

File: nixos/tests/starship.nix
Comment (pennae) [reply nixos/tests/starship.nix:22]: personally wouldn't say this is an unnecessary VM test since it checks that the module has done the correct setup for shells to use the prompt. can that even be done from installCheck?

Comment (danth) [reply nixos/tests/starship.nix:22]: The `expect` script could run in `installCheckPhase`; however that would only test that the package is working, not that the NixOS module has installed it correctly.

Comment (marsam) [reply nixos/tests/starship.nix:22]: I see, It makes sense keeping it as vm test then.

ping me after you resolve the conflicts to merge it :)

**Comment by danth** [comment]
@marsam Sorry, it seems the test code got reverted to an older commit when I rebased. Should I open a new PR to fix or is there anything else you can do?

**Comment by marsam** [comment]
>Should I open a new PR to fix?

yes, please

**Pull request merged**
